### ISSUE TITLE
[Problem] TechDraw: Feature for GD&T or Improvement of the Balloon Workaround

### ISSUE THREAD

**Issue by maxwxyz** [FreeCAD/FreeCAD#11112]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

There is no explicit feature for GD&T (https://en.wikipedia.org/wiki/Geometric_dimensioning_and_tolerancing) in TechDraw.
I know that it would be ideal to implement in the 3D environment and derive the drawing in TechDraw later but there is currently no active or working GD&T Workbench in FreeCAD. https://github.com/5axes/FreeCAD-GDT

As ToniTen mentioned [here ](https://forum.freecad.org/viewtopic.php?t=68639) the current workaround to hack GD&T into rectangular balloons is not ideal and produces results which need manual adjustments of the final .svg drawing in external applications.

If a dedicated GD&T feature for ISO GPS is neglected in FreeCAD, we should at least have a better workaround to be able to use FreeCAD in professional environments:

- Icons and text in rectangular balloons are currently not centeres
- No feature to create multi-line balloons 
- Remove excessive whitespace inside the rectangles
- Make the first rectangle always a square and the same size when using  GD&T symbols in it
- Make the leader line connected to the balloon features
- Include all operators of ISO GPS e.g.:
![ISO7083](https://github.com/FreeCAD/FreeCAD/assets/6246609/26091a0b-c76c-4514-8dcd-c48725bfc013)

### Full version info

```shell
0.21
```

### Subproject(s) affected?

Techdraw

### Anything else?

https://forum.freecad.org/viewtopic.php?f=10&t=22072

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by maxwxyz** [comment]
Most icons are in ISO 1101

**Comment by sliptonic** [comment]
@maxwxyz  there are a handful of issues related to GD&T.  Will you create a project to group them?  Then we can start tracking them together and look at getting some FPA grant money assigned.

**Comment by maxwxyz** [comment]
@sliptonic yes, could you create a project, we need an organization owner for this. If possible, please also one for UI stuff.

**Comment by sliptonic** [comment]
Done.  I've invited you to both as an admin

https://github.com/orgs/FreeCAD/projects/23/views/1
https://github.com/orgs/FreeCAD/projects/24/views/1

**Comment by maxwxyz** [comment]
@sliptonic thanks, will add matching issues and create new ones.
Is it meant for TechDraw only or also in the 3D view? Best way would be support in the 3D view as most applications do to add tolerances to datums on the model and derive the drawing from it. Faster could be the implementation in 2D only I guess.

There was an addon for the 3d implementation but it is not maintained: https://github.com/5axes/FreeCAD-GDT

**Comment by sliptonic** [comment]
Yeah,  GD&T is a bigger topic.  It certainly affects TechDraw but also the 3D window and the Step export facility.  

There's a fork of the addon that works and isn't in the addon manager.   

Basically the FreeCAD community needs to have a bigger discussion about GD&T and how we support it.

**Comment by maxwxyz** [comment]
I guess it depends on the resources. 3D support + step would be the go to for professional applications but easier could be the hack with the balloon feature.

**Comment by lucas-bueno777** [comment]
Hi! I am not sure if this is the best place to comment about this, but I like the idea of adding GD&T in the 3D view, such as Functional tolerancing and Annotation (FTA) for CATIA. It is widely used in the industry - but probably is very hard to implement.

Thanks for the amazing work.

**Comment by maxwxyz** [comment]
@sliptonic there are not more issues directly related to GD&T. I could go through each ISO GPS standard and create separate issues of what is needed to fully implement this. 
But I guess this is a lot of work and will need some time as it's not just a list of what's missing but it has to be grouped and ideally proposed as a functioning UI concept.

I could also try to create the issues so they are for the 3D implementation and the drawing could be derived later, as GD&T should be done professionally. So feature requests for a GD&T WB and TechDraw could grab this info for the drawing.
If this could be implemented in 3D via a new GD&T WB there could be also the possibility for future WBs to build on top using the information for tolerance analyzes / variance analyzes and so on.

Is there any other person who is familiar with GD&T who could support?

This is just the base of standards I've grabbed from an book:
![image](https://github.com/FreeCAD/FreeCAD/assets/6246609/6e496154-adfa-4a05-8f48-b0ea82b4c60f)

**Comment by sliptonic** [comment]
Even the assumption that there should be "a GD&T workbench" may be premature.  For all we know, GD&T functions rightly belong in suitable places throughout the application.  

I don't think we should go and create the whole raft of issues up front.  I'd prefer to have the issues be driven from user-facing problems.  We could hunt through the forum and see if users have been asking for GD&T functionality.  Then tap them to create specific issues in the tracker.

I don't know of any local experts in the field.  Issue #1 might be to identify a project manager to figure out what needs to be done. That person might come from the forum or from another source.  I'll go through my contact information to see if I know anyone.

**Comment by pure-machinery** [comment]
From my side it is best to tackle 2D first, but this would require better logic for annotations or leader lines even before GD&T symbols are added. The symbols can just be used from a .ttf file.

If 3D is tackled first then I would say there should be an ability to pull it into 2D from the model. This would be awesome but I find it hard to believe FC will do something like this. 

Creo has pulling dimensions and annotations from features into the drawing for example. This makes creating drawings very easy and it makes it hard to forget if you've added a specific dimension.

**Comment by cooper-david-a** [comment]
The problem I've always had with GD&T in 3D models is that the vendor with whom you are communicating needs to be able to open the 3D file.

The latest .step files are the only neutral format that I am aware of that will preserve the annotations and I have not been very impressed with the result so far.  I **don't** think the 3D pdf views produced by MBD features provide much actual utility.

Unless you do all of your manufacturing in house or are a large company you have to work with what your vendor uses.  Typically in my case this means providing a 3D model in a CAD neutral format for them to use in their CAM software and a 2D drawing to handle all of the tolerances and notes.

Once you've arrived at the point where 3D annotations are not useful in getting parts produced, there is no point in placing them in the 3D environment at all.

Honestly, the manual method of attaching GD&T symbols I saw here [here](https://youtu.be/7GwaIpJgp1Q?si=vAKHA9P0FGG64uD8) is adequate.  Maybe some minor tweaks like making the symbols look better and allowing feature control frames to be grouped with dimensions (so they move together).

I am actually most interested in a CAD software that will run on Linux.

**Comment by maxwxyz** [comment]
IMO 3D attachment is crucial. Not only for exchanging in STEP format but also the ability to perform tolerance analysis in assemblies. For manufacturing and quality control (manually) the 2D drawing is sufficient. But there are measurement machines out there that can work with 3D models, grad the properties (product specifications) per surface / axis out of the 3D file and derive the right measurement program and compare the sample.
The attachment like annotations in 3D are only interesting for the human and to be able to assign tolerances and check them. The data should be stored as properties for the selected features.

**Comment by cooper-david-a** [comment]
I'm not familiar enough with tolerance stackup or  measurement software to comment on those use cases.  I am curious whether there will not need to be interfaces with a bunch of proprietary file formats or if those software support some neutral format.

It looks like [AP242](http://www.ap242.org/welcome.html) is addressing the interoperability issues I had with MBD last time I looked into it.  If the latest systems are capable of importing the GD&T from a neutral format like .step AP242 in a usable way and not just the annotations, that changes things.

Until 3D data is supported broadly, (which might be happening now) I would just hate to loose the simple and elegant solution of updating GD&T feature control frames right there on the 2D drawing.

**Comment by luzpaz** [comment]
Heads up: seems Ondsel is interested in implementing GD&T https://ondsel.com/blog/gd-and-t-as-core-feature/

**Comment by lucas-bueno777** [comment]
[Here](https://www.wichita.edu/industry_and_defense/NIAR/Documents/Functional-Tolerancing-Annotations-R30.pdf) is a publicly available file that shows how CATIA handles GD&T, in case that helps in the brainstorming.

**Comment by cooper-david-a** [comment]
> Heads up: seems Ondsel is interested in implementing GD&T https://ondsel.com/blog/gd-and-t-as-core-feature/

Yeah, I read that article.

For some background, I am a mechanical engineer and have been working in the aerospace industry for around 13 years now.  I've used Solid Works and SolidEdge extensively and have been recently using Fusion.

I played around with FreeCAD about 10years ago for a year or two.  I just started following it again as I am interested in getting away from Windows.

It could be that I am unusual but when I am developing geometry in CAD I am doing something fundamentally different then when I am making a 2D drawing to send to a manufacturer.  Developing the geometry can be a messy process where it makes sense to define geometry off of mating components or reference planes or other weird stuff that wouldn't be in a 2D drawing.  Making a 2D drawing is specifying a set of manufacturing requirements.  If you are able to do both at the same time, you are much better at this than I am.

I personally haven't found the ability to import dimensions from the 3D geometry very useful at all.  I always start with a blank view and add dimensions and tolerances in the 2D environment.

Progress is apparently being made toward a 3D only manufacturing workflow.  However, I think it is still a long way off for those of us who can't mandate what software our vendors use.

**Comment by Mannshoch** [comment]
@cooper-david-a There are two concepts about messuring GD&T/ISO-GPS one is about following the design features and one is about, easuring in 3D after the design. Your Way of Working is mor about Measuring after the design, but that could also happen in 3D.
I know companies, that which have the rule, that the moment before a part switch from developing to the first release the part has to be newly created from the Ground. Benefit is, that while developing a part you have not to focus on a sustainable feature tree. A redesign is mostly done really fast and while doing you could design it according to needs for variants, and measure it for auto import into Drawings, and [add comment](https://help.solidworks.com/2025/english/solidworks/sldworks/hidd_comment_dlg.htm).

If you work that way you could also add tolerances and GD&T/ISO-GPS symbols.

I assume a combined possibility would be best. Measure it in sketch, and add gd&t there if useful and otherwise add gd&t from external. 

I think, influenced by Solidwork, there must be a possibility to see all GD&T/ISO-GPS symbols used in 3D and modify from there. Like removing/attaching a symbol to a sketch-dimension and/or add/remove a 3D-Measure to the body.

**Comment by cooper-david-a** [comment]
@Mannshoch, Good points.  I may try the method you describe on my next few drawings.  I don't understand the advantages of such an approach at this point but I'll report back.  I'll also discuss this with some of my suppliers.  They may be further along than I am giving them credit for.

**Comment by pure-machinery** [comment]
One thing I would like to add is that this will affect the snap lines implementation in Freecad. 

Currently we have a fixed width between dimensions but what happens if you have a tolerance frame (or multiple) beneath the dimension line? 

You could argue that the tolerance frames could be used when calculating the height of the dimension and just have the same offset between the last tolerance frame and the highest point of the dimension below.

**Comment by pierreporte** [comment]
A **150 € reward** will be given to the contributor that writes a pull request, solving this issue, that gets merged. It will be paid by the FreeCAD Project Association. It is to be considered an incentive. Grant applications are welcome, but the total amount will exclude the reward.

**Comment by maxwxyz** [comment]
This issue is about 2D GD&T support in TechDraw. I have created an overall issue to group sub-issues related to the 3D workflow:
- https://github.com/FreeCAD/FreeCAD/issues/29772

**Comment by mohamed-khairy-swe** [comment]
Can I tackle this issue?
I will follow the check list @maxwxyz  made so that means I will start with the "Icons and text in rectangular balloons are currently not centered" task and then make a PR for it and then move to the next task until the check list is done.

I'm still learning my way around FreeCAD and the codebase so it may take me a while to close this issue , that's why I wanted to check if someone is working on this?

**Comment by maxwxyz** [comment]
Go ahead, I don't know about someone working on this

**Comment by mohamed-khairy-swe** [comment]
Hey @maxwxyz , I have been thinking since this issue is about 2D GD&T support in TechDraw, should I work on fully supporting 2D GD&T in TechDraw instead of just making a balloon workaround and calling it a day?

**Comment by pierreporte** [comment]
The balloon workaround is a workaround. The idea is to have full support of the standard in a sane and robust way.

**Comment by mohamed-khairy-swe** [comment]
@pierreporte oh so does that means fully supporting 2D GD&T in TechDraw through balloon?

**Comment by pierreporte** [comment]
The opposite actually

**Comment by mohamed-khairy-swe** [comment]
@pierreporte I'm lost, so what you're trying to say is that this issue focus on making a workaround and a full support for 2D GD&T in TechDraw needs it's own issue?

**Comment by pierreporte** [comment]
This issue doesn’t focus on a workaround. It asks for the real thing, or at the very least a workaround with balloon. I don’t think it would be resolved until the proper feature is there. But you can still make something with the workaround.

**Comment by mohamed-khairy-swe** [comment]
@pierreporte Oh wow, I thought I'd be closing this issue once the workaround was done but this is going to be a very big task. I'll need to reread the experts' comments on this issue, plus do more research and planning, to come up with the best design for fully supporting 2D GD&T in TechDraw.

Should I stop working on the workaround and focus on full support, or finish the workaround first and then move on to full support?

**Comment by pierreporte** [comment]
At the end if the day it's up to you.

**Comment by mohamed-khairy-swe** [comment]
Ok I will see what I can do, I may write an implementation plan and send it here for you to review it so stay tuned.

**Comment by franklincg** [comment]
Hi, I'm interested in working on this bounty. Could you confirm the current reward amount and whether the issue is still available for a new contributor to work on?

**Comment by mohamed-khairy-swe** [comment]
@franklincg 
sorry but I'm currently working on this issue, but I encourage you to work on this bounty #8076 because it would be a really nice addition to FreeCAD, and because you are a new contributor please don't use AI , try to tackle the issue on your own and plan your approach on how you are going to solve it and read the codebase and really level up your cognitive capabilities.
